presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PushLimitThroughOuterJoin.java

```diff
@@ -0,0 +1,99 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.sql.planner.iterative.rule;
+
+import com.facebook.presto.matching.Capture;
+import com.facebook.presto.matching.Captures;
+import com.facebook.presto.matching.Pattern;
+import com.facebook.presto.sql.planner.iterative.Lookup;
+import com.facebook.presto.sql.planner.iterative.Rule;
+import com.facebook.presto.sql.planner.plan.JoinNode;
+import com.facebook.presto.sql.planner.plan.JoinNode.Type;
+import com.facebook.presto.sql.planner.plan.LimitNode;
+import com.facebook.presto.sql.planner.plan.PlanNode;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.Range;
+
+import static com.facebook.presto.matching.Capture.newCapture;
+import static com.facebook.presto.sql.planner.optimizations.QueryCardinalityUtil.extractCardinality;
+import static com.facebook.presto.sql.planner.plan.JoinNode.Type.FULL;
+import static com.facebook.presto.sql.planner.plan.JoinNode.Type.LEFT;
+import static com.facebook.presto.sql.planner.plan.JoinNode.Type.RIGHT;
+import static com.facebook.presto.sql.planner.plan.Patterns.Join.type;
+import static com.facebook.presto.sql.planner.plan.Patterns.join;
+import static com.facebook.presto.sql.planner.plan.Patterns.limit;
+import static com.facebook.presto.sql.planner.plan.Patterns.source;
+
+public class PushLimitThroughOuterJoin
+        implements Rule<LimitNode>
+{
+    private static final Capture<JoinNode> CHILD = newCapture();
+
+    private static final Pattern<LimitNode> PATTERN =
+            limit()
+                    .with(source().matching(
+                            join()
+                                    .with(type().matching(type -> isLeftOuter(type) || isRightOuter(type)))
+                                    .capturedAs(CHILD)));
+
+    @Override
+    public Pattern<LimitNode> getPattern()
+    {
+        return PATTERN;
+    }
+
+    @Override
+    public Result apply(LimitNode parent, Captures captures, Context context)
+    {
+        JoinNode joinNode = captures.get(CHILD);
+        PlanNode left = joinNode.getLeft();
+        PlanNode right = joinNode.getRight();
+
+        boolean changed = false;
+
+        if (isLeftOuter(joinNode.getType()) && !isLimited(left, context.getLookup(), parent.getCount())) {
+            left = new LimitNode(context.getIdAllocator().getNextId(), left, parent.getCount(), false);
+            changed = true;
+        }
+
+        if (isRightOuter(joinNode.getType()) && !isLimited(right, context.getLookup(), parent.getCount())) {
+            right = new LimitNode(context.getIdAllocator().getNextId(), right, parent.getCount(), false);
+            changed = true;
+        }
+
+        if (changed) {
```

Comment:
As Martin suggested before I would inline this variable and used `left != joinNode.getLeft() || right != joinNode.getRight()`